src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp

```diff
@@ -537,22 +603,48 @@ static ColumnWithTypeAndName readIPv6ColumnFromBinaryData(std::shared_ptr<arrow:
         for (size_t i = 0; i != chunk_length; ++i)
         {
             /// If at least one value size is not 16 bytes, fallback to reading String column and further cast to IPv6.
-            if (chunk.value_length(i) != sizeof(IPv6))
+            if (!chunk.IsNull(i) && chunk.value_length(i) != sizeof(IPv6))
                 return readColumnWithStringData<arrow::BinaryArray>(arrow_column, column_name);
         }
         total_size += chunk_length;
     }
 
     auto internal_type = std::make_shared<DataTypeIPv6>();
     auto internal_column = internal_type->createColumn();
-    auto & data = assert_cast<ColumnIPv6 &>(*internal_column).getData();
-    data.reserve(total_size * sizeof(IPv6));
+    auto & ipv6_column = assert_cast<ColumnIPv6 &>(*internal_column);
+    ipv6_column.reserve(total_size);
 
     for (int chunk_i = 0, num_chunks = arrow_column->num_chunks(); chunk_i < num_chunks; ++chunk_i)
     {
         auto & chunk = dynamic_cast<arrow::BinaryArray &>(*(arrow_column->chunk(chunk_i)));
-        const auto * raw_data = reinterpret_cast<const IPv6 *>(chunk.raw_data() + chunk.raw_value_offsets()[0]);
-        data.insert_assume_reserved(raw_data, raw_data + chunk.length());
+        for (size_t value_i = 0, length = static_cast<size_t>(chunk.length()); value_i < length; ++value_i)
+        {
+            if (chunk.IsNull(value_i))
+                ipv6_column.insertDefault();
+            else
+                ipv6_column.insertData(chunk.Value(value_i).data(), chunk.Value(value_i).size());
+        }
+    }
+    return {std::move(internal_column), std::move(internal_type), column_name};
+}
+
+static ColumnWithTypeAndName readIPv4ColumnWithInt32Data(std::shared_ptr<arrow::ChunkedArray> & arrow_column, const String & column_name)
+{
+    auto internal_type = std::make_shared<DataTypeIPv4>();
+    auto internal_column = internal_type->createColumn();
+    auto & column_data = static_cast<ColumnIPv4 &>(*internal_column).getData();
+    column_data.reserve(arrow_column->length());
+
+    for (int chunk_i = 0, num_chunks = arrow_column->num_chunks(); chunk_i < num_chunks; ++chunk_i)
```

Comment:
?
```suggestion
    for (size_t chunk_i = 0, num_chunks = arrow_column->num_chunks(); chunk_i < num_chunks; ++chunk_i)
```

Reply:
Index `chunk_i` is used in `arrow_column->chunk()` method and it takes `int` argument. We can use `size_t` in loop, but we will need to cast it to `int` explicitly anyway.

Reply:
ah, ok, did not notice, then leave as is.